File: sdk/android/frak-sdk/src/main/kotlin/id/frak/sdk/tracking/EventOutbox.kt
package id.frak.sdk.tracking

import id.frak.sdk.config.Backoff
import id.frak.sdk.core.FrakLogger
import id.frak.sdk.identity.ProofOp
import id.frak.sdk.net.HttpClient
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONObject
import java.util.UUID
import java.util.concurrent.atomic.AtomicInteger

/**
 * Queues tracked events and drains them oldest first: enqueue is durable, sending is best-effort
 * behind it, one send in flight at a time. Delivery for a row's [QueuedRow.kind] is delegated to
 * [senders]; this class only owns ordering, consent, backoff, reconciliation and the failure cap.
 */
internal class EventOutbox(
    private val queue: EventQueue,
    private val http: HttpClient,
    private val logger: FrakLogger,
    /** Drains run here, not on the caller: [track] must not block on the whole backlog. */
    private val scope: CoroutineScope,
    /** The id events are currently captured under, read fresh so a reset is visible mid-drain. */
    private val currentClientId: suspend () -> String?,
    /**
     * Whether tracking is still permitted, re-read per event: [flush] posts outside [queueMutex],
     * so without this an event already read would still upload after a withdrawal.
     */
    private val trackingAllowed: suspend () -> Boolean = { true },
    /**
     * Resolves the merchant for a row captured without one, at drain time. Called at most once
     * per drain and only when a deferred row is actually reached; null means "not yet", which
     * holds the row rather than dropping it. Never let this throw.
     */
    private val resolveMerchantId: suspend () -> String? = { null },
    /** Mints a proof for a sender that needs one (currently only [MergeSender]). Never let this throw. */
    private val signProof: suspend (
        op: ProofOp,
        merchantId: String,
        binding: ByteArray,
    ) -> String? = { _, _, _ -> null },
    /** One [RowSender] per [QueuedRow.kind]; an unrecognised kind is skipped, never generic-POSTed. */
    private val senders: Map<String, RowSender> = emptyMap(),
    private val now: () -> Long = System::currentTimeMillis,
    private val newKey: () -> String = { UUID.randomUUID().toString() },
    private val backoff: Backoff = Backoff(now),
) {
    /** Guards the file. Held only for reads and writes, never across a request. */
    private val queueMutex = Mutex()

    /** One drain at a time. A second concurrent flush would reorder the queue it is draining. */
    private val flushMutex = Mutex()

    /** [IDLE] / [DRAINING] / [DRAIN_AGAIN]; see [scheduleFlush]. */
    private val drainState = AtomicInteger(IDLE)

    /**
     * Returns once the event is durable, not once delivered: the drain is detached onto [scope].
     *
     * A null [merchantId] is captured anyway and resolved at drain ([resolveMerchantId]), so a
     * referral arriving before any config could be resolved is held rather than lost.
     */
    suspend fun track(
        merchantId: String?,
        clientId: String?,
        interaction: Interaction,
    ) {
        val key = (interaction.kind as? Interaction.Kind.Custom)?.idempotencyKey ?: newKey()
        enqueue(InteractionSender.KIND, interactionPayload(interaction, key), clientId, merchantId, key)
        scheduleFlush()
    }

    /** Same contract as [track]. Idempotency key never reaches the wire; backend dedupes on `(orderId, token)`. */
    suspend fun trackPurchase(
        merchantId: String?,
        clientId: String?,
        customerId: String,
        orderId: String,
        token: String,
    ) {
        val payload =
            JSONObject()
                .put("customerId", customerId)
                .put("orderId", orderId)
                .put("token", token)
        enqueue(PurchaseSender.KIND, payload, clientId, merchantId, newKey())
        scheduleFlush()
    }

    // No shutdown(): every drain is launched on the client's scope, which
    // DefaultFrakClient.shutdown() cancels and joins.

    /**
     * Collapses a burst of enqueues into one drain, mirroring iOS's `drainTask`/`drainAgain`: a
     * request arriving while a drain runs re-runs it once at the end, instead of launching one
     * coroutine per event that then serialise on [flushMutex] and re-read the file each time.
     *
     * A cancellation leaves this at [DRAINING] — the only canceller is `shutdown()`, after which
     * nothing enqueues again.
     */
    private fun scheduleFlush() {
        while (true) {
            when (drainState.get()) {
                IDLE -> {
                    if (drainState.compareAndSet(IDLE, DRAINING)) {
                        scope.launch {
                            try {
                                do {
                                    drainState.set(DRAINING)
                                    flush()
                                } while (!drainState.compareAndSet(DRAINING, IDLE))
                            } catch (failure: Throwable) {
                                // Reset before rethrowing: the client's handler logs this and the
                                // process survives, and a state left at DRAINING would turn every
                                // later enqueue into a silent no-op.
                                drainState.set(IDLE)
                                throw failure
                            }
                        }
                        return
                    }
                }

                DRAINING -> {
                    if (drainState.compareAndSet(DRAINING, DRAIN_AGAIN)) return
                }

                else -> {
                    return
                }
            }
        }
    }

    /** Called on anonymous id reset: an event captured under the dead id must never be emitted. */
    suspend fun purge() {
        queueMutex.withLock { queue.clear() }
    }

    /**
     * Queues an inbound identity merge, with the same durability contract as [track]: an `fmt`
     * token is single-use and short-lived, so losing one to a cold cache or a dead network is
     * permanent, and it must not depend on a merchant resolving before it can be written down.
     */
    suspend fun trackMerge(
        merchantId: String?,
        anonymousId: String,
        mergeToken: String,
    ) {
        // A token already on disk was claimed by a previous process, whose in-memory guard died
        // with it; without this, every cold start on the same intent would queue it again.
        if (isQueued(MergeSender.KIND, mergeToken)) return
        enqueue(MergeSender.KIND, JSONObject(), anonymousId, merchantId, mergeToken)
        scheduleFlush()
    }

    /** [idempotencyKey] is the token itself for a merge row; it never reaches the wire from there. */
    private suspend fun isQueued(
        kind: String,
        idempotencyKey: String,
    ): Boolean =
        queueMutex
            .withLock { queue.read(now()) }
            .any { it.kind == kind && it.idempotencyKey == idempotencyKey }

    /**
     * Books one failed delivery against [event]: retried until [MAX_FAILURES], then dropped so a
     * permanently-rejected row cannot block the FIFO behind it forever.
     */
    private fun recordRetry(
        event: QueuedRow,
        delivered: MutableSet<Long>,
        retried: MutableMap<Long, QueuedRow>,
        reason: String,
    ) {
        val failed = event.withFailure()
        if (failed.failures >= MAX_FAILURES) {
            logger.warn("Dropping an event: $reason.")
            delivered += event.rowId
        } else {
            retried[event.rowId] = failed
        }
    }

    /**
     * Sends oldest first, stopping at a transient failure or a hold to keep FIFO order, and
     * skipping ahead past anything whose verdict concerns that row alone: a rejection, an
     * unrecognised kind, one already delivered. Network I/O happens outside [queueMutex], and
     * the file is reconciled against a fresh read so a mid-flush append isn't lost.
     */
    suspend fun flush() {
        flushMutex.withLock {
            if (!trackingAllowed()) return
            if (backoff.isBackingOff(BACKOFF_KEY)) return

            val pending = queueMutex.withLock { queue.read(now()) }
            if (pending.isEmpty()) {
                // Expired or unreadable rows may still be on disk.
                queueMutex.withLock { queue.reconcile(emptySet(), emptyMap(), now()) }
                return
            }

            val currentClientId = currentClientId()
            val delivered = mutableSetOf<Long>()
            val retried = mutableMapOf<Long, QueuedRow>()
            val ctx = sendContext()

            // A kill mid-drain replays whatever this pass has sent but not yet written down, so
            // the accumulator is checkpointed rather than held for the whole backlog.
            suspend fun checkpoint() {
                if (delivered.isEmpty() && retried.isEmpty()) return
                queueMutex.withLock { queue.reconcile(delivered.toSet(), retried.toMap(), now()) }
                delivered.clear()
                retried.clear()
            }

            for ((index, event) in pending.withIndex()) {
                if (index > 0 && index % RECONCILE_EVERY == 0) checkpoint()
                // break, not return: falling through to reconcile prevents a re-send when a
                // concurrent purge's file delete silently fails.
                if (!trackingAllowed()) {
                    logger.info("Tracking was disabled mid-drain; stopping without sending the rest.")
                    break
                }
                // Dropped even if purge and this drain raced: event carries the id it was captured under.
                if (event.clientId != null && currentClientId != null && event.clientId != currentClientId) {
                    delivered += event.rowId
                    continue
                }

                val sender = senders[event.kind]
                if (sender == null) {
                    logger.warn("No sender registered for row kind '${event.kind}'; skipping it.")
                    continue
                }

                // Stamped at drain, not left null: capture deliberately does not gate on an id
                // (a locked device has none yet), and a row sent without `x-frak-client-id` is a
                // guaranteed 401 that would spend the failure cap and land unattributed.
                val row =
                    if (event.clientId == null &&
                        currentClientId != null
                    ) {
                        event.withClientId(currentClientId)
                    } else {
                        event
                    }

                val outcome =
                    if (row.clientId == null) DeliveryOutcome.Hold else sender.deliver(row, ctx)
                when (outcome) {
                    is DeliveryOutcome.Delivered -> {
                        backoff.recordSuccess(BACKOFF_KEY)
                        delivered += row.rowId
                        continue
                    }

                    is DeliveryOutcome.Dropped -> {
                        delivered += row.rowId
                        continue
                    }

                    is DeliveryOutcome.Retryable -> {
                        backoff.recordFailure(BACKOFF_KEY, outcome.error)
                        break
                    }

                    is DeliveryOutcome.Rejected -> {
                        // continue, not break: a rejection is a verdict on this row alone, and
                        // one poison row must not stall every event queued behind it.
                        recordRetry(row, delivered, retried, "the backend rejected it")
                        continue
                    }

                    is DeliveryOutcome.Hold -> {
                        if (row.heldSince == null) {
                            retried[row.rowId] = row.withHeldSince(now())
                            break
                        }
                        val heldFor = now() - row.heldSince
                        if (heldFor <= sender.holdTimeoutMillis) break
                        // continue, not break: dropping the dead row is the point, and the rows
                        // behind it must get their turn in this same drain.
                        logger.warn("Dropping a '${row.kind}' row held ${heldFor}ms with no resolvable inputs.")
                        delivered += row.rowId
                        continue
                    }
                }
            }

            // One EventQueue-owned hop: a read()-then-replace() from here would erase an event
            // appended between the two suspending calls. Unconditional even when the accumulator
            // is empty: expired or unreadable rows may still be on disk.
            queueMutex.withLock {
                queue.reconcile(delivered, retried, now())
            }
        }
    }

    /**
     * Built once per drain, so [SendContext.resolveMerchantId] resolves at most once however many
     * deferred rows this pass reaches — matching [resolveMerchantId]'s own once-per-drain contract.
     */
    private fun sendContext(): SendContext {
        var merchantId: String? = null
        var attempted = false
        return SendContext(
            http = http,
            resolveMerchantId = {
                if (!attempted) {
                    attempted = true
                    merchantId = resolveMerchantId()
                }
                merchantId
            },
            signProof = signProof,
        )
    }

    private suspend fun enqueue(
        kind: String,
        payload: JSONObject,
        clientId: String?,
        merchantId: String?,
        idempotencyKey: String,
    ) {
        queueMutex.withLock {
            // EventQueue.append assigns and persists the real row id; no caller may choose one.
            queue.append(
                QueuedRow(
                    idempotencyKey = idempotencyKey,
                    kind = kind,
                    payload = payload,
                    clientId = clientId,
                    merchantId = merchantId,
                    capturedAtMillis = now(),
                    rowId = EventQueue.MISSING_ROW_ID,
                ),
            )
        }
    }

    /** Idempotency key is only written for the two shapes whose schema carries one; `arrival` has none. */
    private fun interactionPayload(
        interaction: Interaction,
        idempotencyKey: String,
    ): JSONObject =
        when (val kind = interaction.kind) {
            is Interaction.Kind.Arrival -> {
                JSONObject()
                    .put("type", "arrival")
                    .put("referrerWallet", kind.referrerWallet)
                    .put("referrerClientId", kind.referrerClientId)
                    .put("referrerMerchantId", kind.referrerMerchantId)
                    .put("referralTimestamp", kind.referralTimestamp)
            }

            is Interaction.Kind.Sharing -> {
                JSONObject()
                    .put("type", "sharing")
                    .put("sharingTimestamp", kind.sharingTimestamp ?: (now() / 1000))
                    .put("purchaseId", kind.purchaseId)
                    .put("idempotencyKey", idempotencyKey)
            }

            is Interaction.Kind.Custom -> {
                // Shape is not validated here — the route's schema is the authority and a
                // rejection is a 4xx — but size is: an unbounded map lands verbatim in a durable
                // file whose only other cap is a row count.
                val data = JSONObject()
                kind.data.entries.take(MAX_CUSTOM_ENTRIES).forEach { (key, value) ->
                    data.put(key.take(MAX_CUSTOM_FIELD_LENGTH), value.take(MAX_CUSTOM_FIELD_LENGTH))
                }
                if (kind.data.size > MAX_CUSTOM_ENTRIES) {
                    logger.warn(
                        "A custom interaction carried ${kind.data.size} data entries; " +
                            "only the first $MAX_CUSTOM_ENTRIES are queued.",
                    )
                }
                JSONObject()
                    .put("type", "custom")
                    .put("customType", kind.customType.take(MAX_CUSTOM_FIELD_LENGTH))
                    .put("data", data)
                    .put("idempotencyKey", idempotencyKey)
            }
        }

    private companion object {
        const val MAX_FAILURES = 3
        const val BACKOFF_KEY = "track"

        /** Rows between mid-drain reconciles. Bounds a duplicate replay to this many rows, at one file rewrite each. */
        const val RECONCILE_EVERY = 20

        /** Bounds on a custom interaction's payload, so one call cannot fill the queue file on its own. */
        const val MAX_CUSTOM_ENTRIES = 32
        const val MAX_CUSTOM_FIELD_LENGTH = 512

        const val IDLE = 0
        const val DRAINING = 1
        const val DRAIN_AGAIN = 2
    }
}

File: sdk/ios/Sources/FrakSDK/Tracking/EventOutbox.swift
import Foundation

/// Captures interactions, purchases and identity merges durably, then drains them oldest-first
/// through the `RowSender` registered for each row's `kind`.
///
/// The queue is the contract: `track` returns once the event is on disk, not once it is
/// delivered. Delivery is a best-effort drain that stops at the first failure, so a later
/// event is never sent before an earlier one has gone through.
actor EventOutbox {
    private static let interactionKind = InteractionSender.kind
    private static let purchaseKind = PurchaseSender.kind
    private static let mergeKind = MergeSender.kind
    /// Permanent rejections tolerated before a row is dropped rather than left blocking
    /// the head of the queue forever.
    private static let maxFailures = 3
    private static let backoffKey = "track"

    private let queue: EventQueue
    private let http: HTTPClient
    private let logger: FrakLogger
    /// The only place a `QueuedRow.kind` is turned into a request. An unregistered kind is
    /// `continue`d in `drain()`, not dropped — see `QueuedRow.currentSchemaVersion`.
    private let senders: [String: any RowSender]
    /// Read fresh on every drain, so an identity reset takes effect mid-flight. Async:
    /// `AnonymousIdStore.anonymousId()` awaits eager generation rather than blocking, so this
    /// drain loop does not call a blocking keystore read from inside itself.
    private let currentClientId: @Sendable () async -> String?
    /// Read fresh inside the drain loop, so a withdrawal mid-drain stops the upload. `purge()`
    /// alone cannot: `drain()` reads the backlog up front, and the stale-id guard below is
    /// disabled rather than tightened by the nil id a withdrawal produces.
    ///
    /// Defaults to always-allowed for this file's own tests; the real gate is wired in
    /// `DefaultFrakClient`.
    private let trackingAllowed: @Sendable () async -> Bool
    /// False while the identity store exists but cannot be read — only before a device's first
    /// unlock. Gates the drain, never capture: a row captured in that window carries no client id
    /// and lands unattributed, where gating capture too would drop it entirely.
    private let identityReadable: @Sendable () async -> Bool
    private let resolveMerchantId: @Sendable () async -> String?
    private let signProof: @Sendable (ProofOp, String, Data) async -> String?
    private let now: @Sendable () -> Date
    private let newKey: @Sendable () -> String

    private var backoff: Backoff
    private var drainTask: Task<Void, Never>?
    /// Set when the queue changes under a running drain, so it loops once more.
    private var drainAgain = false
    /// Bumped every time `drainTask` is replaced or cleared, so a finishing drain can tell
    /// whether the slot still holds it before nilling it out.
    private var drainToken = 0
    /// Set once by `shutdown()`, never cleared: a torn-down outbox starts no further drains.
    private var stopped = false
    /// Memoised across one drain's rows. Outer nil: not attempted. Inner nil: attempted, nothing
    /// resolved. Actor state, not a captured `var`, which Swift 6 forbids in a `@Sendable`.
    private var drainMerchantId: String??

    init(
        queue: EventQueue,
        http: HTTPClient,
        logger: FrakLogger,
        senders: [String: any RowSender],
        currentClientId: @escaping @Sendable () async -> String?,
        resolveMerchantId: @escaping @Sendable () async -> String?,
        signProof: @escaping @Sendable (ProofOp, String, Data) async -> String?,
        trackingAllowed: @escaping @Sendable () async -> Bool = { true },
        identityReadable: @escaping @Sendable () async -> Bool = { true },
        now: @escaping @Sendable () -> Date = { Date() },
        newKey: @escaping @Sendable () -> String = { UUID().uuidString.lowercased() },
        backoff: Backoff = Backoff()
    ) {
        self.queue = queue
        self.http = http
        self.logger = logger
        self.senders = senders
        self.currentClientId = currentClientId
        self.resolveMerchantId = resolveMerchantId
        self.signProof = signProof
        self.trackingAllowed = trackingAllowed
        self.identityReadable = identityReadable
        self.now = now
        self.newKey = newKey
        self.backoff = backoff
    }

    func track(merchantId: String?, clientId: String?, interaction: Interaction) async {
        let key = idempotencyKey(for: interaction)
        let body = interactionBody(interaction: interaction, idempotencyKey: key)
        await enqueue(
            kind: Self.interactionKind,
            payload: body,
            clientId: clientId,
            merchantId: merchantId,
            idempotencyKey: key
        )
        detachDrain()
    }

    func trackPurchase(
        merchantId: String?,
        clientId: String?,
        customerId: String,
        orderId: String,
        token: String
    ) async {
        // No idempotency key: the purchase route carries no such field and reconciles on
        // `(orderId, token)` server-side.
        let body = Self.json([
            "customerId": customerId,
            "orderId": orderId,
            "token": token,
        ])
        await enqueue(
            kind: Self.purchaseKind,
            payload: body,
            clientId: clientId,
            merchantId: merchantId,
            idempotencyKey: newKey()
        )
        detachDrain()
    }

    /// Enqueues an identity merge. `clientId` MUST be the anonymousId being folded in: it is
    /// what lets the drain's stale-id guard drop this row after an identity reset instead of
    /// posting a merge for an identity the caller already walked away from.
    func trackMerge(mergeToken: String, anonymousId: String, merchantId: String?) async {
        // A token already on disk was claimed by a previous process, whose in-memory guard died
        // with it; without this, every cold start on the same intent would queue it again.
        // Checked here rather than by the caller so the check and the append are one hop into
        // this actor: two of them would let interleaved calls both pass before either appends.
        guard await !isQueued(kind: Self.mergeKind, idempotencyKey: mergeToken) else { return }
        await enqueue(
            kind: Self.mergeKind,
            payload: "{}",
            clientId: anonymousId,
            merchantId: merchantId,
            idempotencyKey: mergeToken
        )
        detachDrain()
    }

    /// [idempotencyKey] is the token itself for a merge row; it never reaches the wire from there.
    private func isQueued(kind: String, idempotencyKey: String) async -> Bool {
        await queue.read(now: now()).contains { $0.kind == kind && $0.idempotencyKey == idempotencyKey }
    }

    /// Starts a drain without waiting for it: `track` is durable once enqueued, and a caller
    /// on a button handler must not block on a whole backlog.
    private func detachDrain() {
        _ = scheduleDrain()
    }

    /// Drops every queued event. An event captured under an id the user asked to be
    /// forgotten must never be emitted.
    func purge() async {
        await queue.clear()
    }

    /// Drains the queue. Awaiting this awaits the drain, including one already under way.
    func flush() async {
        await scheduleDrain().value
    }

    /// Cancels an in-flight drain and refuses to start another, so `DefaultFrakClient.shutdown()`
    /// leaves nothing running and nothing able to start. One-way: a torn-down outbox is dead.
    ///
    /// The refusal is the load-bearing half — `scheduleDrain` uses `Task.init`, which does not
    /// inherit cancellation, so the next `track()` would start a fresh, uncancelled drain. Does
    /// not await the cancelled task; queued events survive on disk, since this is not an erasure.
    func shutdown() {
        stopped = true
        drainToken += 1
        drainTask?.cancel()
        drainTask = nil
        drainAgain = false
    }

    /// Returns the drain covering everything enqueued so far. A drain already under way is
    /// reused rather than followed by a second full pass: it re-reads the file, so noting that
    /// the queue changed is enough. Awaiting the returned task therefore awaits a pass that
    /// includes the caller's own event.
    private func scheduleDrain() -> Task<Void, Never> {
        // Returns an already-finished task rather than nil, so `flush()`'s `await …value` still
        // has something to await and every caller keeps one shape. See `shutdown()`.
        guard !stopped else { return Task {} }
        if let inFlight = drainTask {
            drainAgain = true
            return inFlight
        }
        // Mirrors `AnonymousIdStore.generationToken`: the tail has to know whether `drainTask`
        // still holds THIS task, and a closure cannot capture the `let` being assigned to it.
        //
        // Without it, a `track()` racing `shutdown()` ends up with two concurrent drains over one
        // queue file — two writers, which `EventQueue` does not serialise.
        drainToken += 1
        let token = drainToken
        let task = Task {
            repeat {
                self.drainAgain = false
                await self.drain()
            } while self.drainAgain && !Task.isCancelled
            if self.drainToken == token { self.drainTask = nil }
        }
        drainTask = task
        return task
    }

    /// A `SendContext` for one drain, with `resolveMerchantId` memoised for that drain's
    /// lifetime. `[weak self]`: a strong capture here would be a retain cycle through
    /// `drainTask`, and the memo itself must live on the actor, not in the closure.
    private func makeSendContext() -> SendContext {
        drainMerchantId = nil
        return SendContext(
            http: http,
            resolveMerchantId: { [weak self] in
                guard let self else { return nil }
                return await self.memoizedMerchantId()
            },
            signProof: signProof
        )
    }

    private func memoizedMerchantId() async -> String? {
        // `if let` because `??` is an autoclosure and rejects `await` on its right-hand side.
        if let cached = drainMerchantId { return cached }
        let resolved = await resolveMerchantId()
        drainMerchantId = resolved
        return resolved
    }

    private func drain() async {
        // Checked before the read, so a drain started by a `track()` that raced a consent
        // withdrawal never even loads the backlog. The per-event re-read below is what closes
        // the withdrawal-lands-mid-drain window; this one just avoids the pointless work.
        guard await trackingAllowed() else { return }
        // Returns before the read, leaving the file as found. An unreadable identity store yields
        // a nil id, which disables the stale-id guard below rather than tightening it.
        guard await identityReadable() else { return }
        guard !backoff.isBackingOff(Self.backoffKey) else { return }

        let pending = await queue.read(now: now())
        guard !pending.isEmpty else {
            // Nothing to send, but expired rows may still be on disk. `reconcile` rather than
            // `replace([])`: read and write must be one hop, or a `track` appending between
            // them is read by neither and erased by the second.
            await queue.reconcile(delivered: [], retried: [:], now: now())
            return
        }

        let currentId = await currentClientId()
        let ctx = makeSendContext()
        var delivered: Set<Int64> = []
        var retried: [Int64: QueuedRow] = [:]

        eventLoop: for event in pending {
            // `break`, not `return`: the caller that flipped consent is about to `purge()`, but a
            // purge that does not land leaves every event this drain already uploaded on disk with
            // no record it went, and `Interaction.arrival` carries no idempotency key — a
            // duplicated referral payout. The cancellation path below must `return` instead.
            guard await trackingAllowed() else {
                logger.info("Tracking was disabled mid-drain; stopping without sending the rest.")
                break eventLoop
            }

            // Every event `queue.read` returns has already been migrated to a non-nil `rowId`;
            // this guard exists so a future caller of `drain()` with a hand-built list can't
            // silently reconcile the wrong row instead of crashing loudly in debug.
            guard let rowId = event.rowId else {
                assertionFailure("a queued event reaching drain() must already have a rowId")
                continue
            }

            // Captured under an id that has since been replaced. Dropped, not sent: it would
            // re-link an identity the user already walked away from.
            if let captured = event.clientId, let currentId, captured != currentId {
                delivered.insert(rowId)
                continue
            }

            guard let sender = senders[event.kind] else { continue }

            let outcome: DeliveryOutcome
            do {
                outcome = try await sender.deliver(row: event, ctx: ctx)
            } catch {
                // Cancellation only — `deliver` is typed `throws(CancellationError)`. Returning
                // skips the reconcile: compacting away an undelivered event is unrecoverable.
                return
            }

            switch outcome {
            case .delivered:
                backoff.recordSuccess(Self.backoffKey)
                delivered.insert(rowId)
            case .dropped:
                delivered.insert(rowId)
            case .retryable(let error):
                backoff.recordFailure(Self.backoffKey, from: error)
                break eventLoop
            case .rejected:
                let failed = event.withFailure()
                if failed.failures >= Self.maxFailures {
                    logger.warn("Dropping a row the backend keeps rejecting (kind \(event.kind)).")
                    delivered.insert(rowId)
                } else {
                    retried[rowId] = failed
                }
                break eventLoop
            case .hold:
                // heldSince is never cleared once set, so this measures total time stuck, not
                // time since the row was last attempted.
                if let heldSince = event.heldSince {
                    guard now().timeIntervalSince(heldSince) > sender.holdTimeout else {
                        break eventLoop
                    }
                    logger.warn(
                        "Dropping a row held past its budget (kind \(event.kind), held \(Int(now().timeIntervalSince(heldSince)))s)."
                    )
                    delivered.insert(rowId)
                    continue eventLoop
                }
                retried[rowId] = event.withHeldSince(now())
                break eventLoop
            }
        }

        // Re-read rather than write back `pending`: a `track` that landed while the drain was
        // awaiting the network is in the file now, and must not be compacted away.
        await queue.reconcile(delivered: delivered, retried: retried, now: now())
    }

    private func enqueue(
        kind: String,
        payload: String,
        clientId: String?,
        merchantId: String?,
        idempotencyKey: String
    ) async {
        await queue.append(
            QueuedRow(
                idempotencyKey: idempotencyKey,
                kind: kind,
                payload: payload,
                clientId: clientId,
                merchantId: merchantId,
                capturedAt: now()
            )
        )
    }

    private func idempotencyKey(for interaction: Interaction) -> String {
        if case .custom(_, _, let supplied) = interaction.kind, let supplied {
            return supplied
        }
        return newKey()
    }

    /// `merchantId` is deliberately absent: a `RowSender` injects it at send time from
    /// `QueuedRow.merchantId`, which is what lets a row be captured before any merchant is known.
    private func interactionBody(interaction: Interaction, idempotencyKey: String) -> String {
        switch interaction.kind {
        case .arrival(let wallet, let clientId, let referrerMerchantId, let timestamp):
            return Self.json([
                "type": "arrival",
                "referrerWallet": wallet,
                "referrerClientId": clientId,
                "referrerMerchantId": referrerMerchantId,
                "referralTimestamp": timestamp,
            ])
        case .sharing(let timestamp, let purchaseId):
            return Self.json([
                "type": "sharing",
                // Unix seconds, stamped now and stored: a retry must not restamp it.
                "sharingTimestamp": timestamp ?? Int64(now().timeIntervalSince1970),
                "purchaseId": purchaseId,
                "idempotencyKey": idempotencyKey,
            ])
        case .custom(let customType, let data, _):
            return Self.json([
                "type": "custom",
                "customType": customType,
                "data": data,
                "idempotencyKey": idempotencyKey,
            ])
        }
    }

    /// Sorted keys so a body is byte-identical every time it is built; the queue stores it
    /// verbatim. Nil-valued keys are absent rather than JSON null, matching the Kotlin twin.
    ///
    /// Every value here is a `String`, `Int64` or `[String: String]`, so the fallback is
    /// unreachable — a fallback rather than a trap, since an SDK does not get to bring down
    /// its host over a body it failed to build.
    private static func json(_ fields: [String: Any?]) -> String {
        let object = fields.compactMapValues { $0 }
        guard let data = try? JSONSerialization.data(withJSONObject: object, options: [.sortedKeys]),
            let text = String(data: data, encoding: .utf8)
        else {
            return "{}"
        }
        return text
    }
}
